File: ros2_ws/src/ugv_bringup_py/ugv_bringup_py/serial_node.py
# This node interacts with the base via UART.
import json

from numpy import deg2rad
import serial

from ament_index_python.packages import get_package_share_directory
from geometry_msgs.msg import Twist
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from rclpy.utilities import ok
from sensor_msgs.msg import Imu, JointState, MagneticField
from std_msgs.msg import Float32, Float32MultiArray

from .utils import BaseController


class SerialNode(Node):
    def __init__(self):
        super().__init__("serial_node")
        # declare parameters
        self.baudrate = int(self.declare_parameter("baudrate", 115200).value or 115200)
        if not self.baudrate:
            self.baudrate = 115200
        self.port = str(self.declare_parameter("port", "/dev/ttyTHS1").value)

        self.pub_name_imu = str(
            self.declare_parameter("pub_name_imu", "imu/data_raw").value
        )
        self.pub_name_mag = str(self.declare_parameter("pub_name_mag", "imu/mag").value)
        self.pub_name_odomraw = str(
            self.declare_parameter("pub_name_odomraw", "encoder").value
        )
        self.pub_name_vol = str(self.declare_parameter("pub_name_vol", "voltage").value)
        self.pub_name_joints = str(
            self.declare_parameter("pub_name_joints", "joint_states").value
        )

        self.sub_name_vel = str(self.declare_parameter("sub_name_vel", "cmd_vel").value)
        self.sub_name_lights = str(
            self.declare_parameter("sub_name_lights", "ugv/led_strl").value
        )
        self.sub_name_servos = str(
            self.declare_parameter("sub_name_servos", "ugv/servos").value
        )

        self.do_servo_calib = bool(self.declare_parameter("do_servo_calib", True).value)
        self.calib_timeout_sec = float(
            self.declare_parameter("calib_timeout_sec", 10.0).value or 10.0
        )
        if self.calib_timeout_sec <= 0:
            self.calib_timeout_sec = 10.0
        sample_period_ms = float(
            self.declare_parameter("sample_period_ms", 50).value or 50.0
        )
        if sample_period_ms <= 0:
            sample_period_ms = 50.0
        self.base = None
        self.calib_pending = False
        # serial setup
        try:
            self.base = BaseController(self.port, self.baudrate, self.get_logger())
            self.base.send_command({"T": 131, "cmd": 1})  # request base info
            # set ESP32 sample rate
            self.base.send_command({"T": 142, "cmd": sample_period_ms})
        except (serial.SerialException, OSError) as e:
            self.get_logger().error(f"Failed to open serial port {self.port}: {e}")
        # sensor data
        self.imu_acce_cov = [0.0] * 9
        self.imu_gyro_cov = [0.0] * 9
        self.mag_cov = [0.0] * 9
        try:
            cov_path = (
                get_package_share_directory("ugv_bringup_py")
                + "/config/covariances.json"
            )
            with open(cov_path, "r", encoding="utf-8") as cov_file:
                covs = json.load(cov_file)
            self.imu_acce_cov = covs.get("imu_acce_covariance", self.imu_acce_cov)
            self.imu_gyro_cov = covs.get("imu_gyro_covariance", self.imu_gyro_cov)
            self.mag_cov = covs.get("magnetic_field_covariance", self.mag_cov)
        except (OSError, json.JSONDecodeError) as e:
            self.get_logger().error(f"Failed to load covariances: {e}")
        # perform servo calibration if required
        if self.do_servo_calib:
            self.start_servo_calib()
        # ROS2 components
        # pubs
        self.pub_imu = self.create_publisher(Imu, self.pub_name_imu, 10)
        self.pub_mag = self.create_publisher(MagneticField, self.pub_name_mag, 10)
        self.pub_odomraw = self.create_publisher(
            Float32MultiArray, self.pub_name_odomraw, 10
        )
        self.pub_vol = self.create_publisher(Float32, self.pub_name_vol, 10)
        self.pub_joints = self.create_publisher(JointState, self.pub_name_joints, 10)
        self.timer = self.create_timer(
            sample_period_ms / 2000.0, self.timer_callback
        )  # set timer period to half of sample period to ensure no data loss
        # subs
        self.sub_vel = self.create_subscription(
            Twist, self.sub_name_vel, self.vel_callback, 1
        )  # TO ensure real time control, set QOS profile to 1
        self.sub_lights = self.create_subscription(
            Float32MultiArray, self.sub_name_lights, self.lights_callback, 1
        )
        self.sub_servos = self.create_subscription(
            Float32MultiArray, self.sub_name_servos, self.servos_callback, 1
        )
        # end of initialization
        self.get_logger().info("Serial Node has been started.")

    def start_servo_calib(self):
        if not self.base:
            self.get_logger().warn("Skipping servo calibration: serial not available.")
            return
        if self.calib_pending:
            return
        self.get_logger().info(
            "Starting servo calibration. Center servos now; will finalize after timeout."
        )
        self.base.send_command({"T": 210, "cmd": 0})  # shutdown servos' torque lock
        self.calib_pending = True
        self.calib_timer = self.create_timer(
            self.calib_timeout_sec, self.finish_servo_calib
        )

    def finish_servo_calib(self):
        if not self.base:
            self.get_logger().warn(
                "Skipping servo calibration finalize: serial not available."
            )
            return
        if not self.calib_pending:
            return
        self.base.send_command({"T": 502, "id": 1})
        self.base.send_command(
            {"T": 502, "id": 2}
        )  # set current position as zero position
        self.calib_pending = False
        self.calib_timer.cancel()
        self.get_logger().info("Servo calibration completed.")

    def timer_callback(self):
        if not self.base:
            return
        base_data = self.base.feedback_data()
        if not base_data:
            return
        if base_data.get("T") == 1001:
            time_now = self.get_clock().now().to_msg()
            self.publish_imu(time_now, base_data)
            self.publish_mag(time_now, base_data)
            self.publish_odomraw(time_now, base_data)
            self.publish_voltage(time_now, base_data)
            self.publish_joints(time_now, base_data)

    def publish_imu(self, time_now, base_data):
        imu_msg = Imu()
        imu_msg.header.stamp = time_now
        imu_msg.header.frame_id = "base_imu_link"

        imu_msg.linear_acceleration.x = 9.8 * float(base_data.get("ax", 0.0)) / 8192
        imu_msg.linear_acceleration.y = 9.8 * float(base_data.get("ay", 0.0)) / 8192
        imu_msg.linear_acceleration.z = 9.8 * float(base_data.get("az", 0.0)) / 8192
        imu_msg.linear_acceleration_covariance = self.imu_acce_cov
        imu_msg.angular_velocity.x = deg2rad(float(base_data.get("gx", 0.0)) / 16.4)
        imu_msg.angular_velocity.y = deg2rad(float(base_data.get("gy", 0.0)) / 16.4)
        imu_msg.angular_velocity.z = deg2rad(float(base_data.get("gz", 0.0)) / 16.4)
        imu_msg.angular_velocity_covariance = self.imu_gyro_cov
        imu_msg.orientation_covariance[0] = -1  # orientation and its covariance not provided

        self.pub_imu.publish(imu_msg)

    def publish_mag(self, time_now, base_data):
        mag_msg = MagneticField()
        mag_msg.header.stamp = time_now
        mag_msg.header.frame_id = "base_imu_link"

        mag_msg.magnetic_field.x = float(base_data.get("mx", 0.0)) * 0.15 * 1e-7
        mag_msg.magnetic_field.y = float(base_data.get("my", 0.0)) * 0.15 * 1e-7
        mag_msg.magnetic_field.z = float(base_data.get("mz", 0.0)) * 0.15 * 1e-7
        mag_msg.magnetic_field_covariance = self.mag_cov

        self.pub_mag.publish(mag_msg)

    def publish_odomraw(self, _, base_data):
        odomraw_msg = Float32MultiArray()
        odomraw_msg.data = [
            float(base_data.get("odl", 0.0)) / 100,
            float(base_data.get("odr", 0.0)) / 100,
            float(base_data.get("L", 0.0)),
            float(base_data.get("R", 0.0)),
        ]
        self.pub_odomraw.publish(odomraw_msg)

    def publish_voltage(self, _, base_data):
        vol_msg = Float32()
        vol_msg.data = float(base_data.get("v", 0.0)) / 100
        self.pub_vol.publish(vol_msg)

    def publish_joints(self, time_now, base_data):
        joints_msg = JointState()
        joints_msg.header.stamp = time_now
        joints_msg.name = ["pt_base_link_to_pt_link1", "pt_link1_to_pt_link2"]
        joints_msg.position = [
            deg2rad(base_data.get("pan", 0.0)),
            deg2rad(base_data.get("tilt", 0.0)),
        ]
        self.pub_joints.publish(joints_msg)

    def vel_callback(self, msg):
        if not self.base:
            return
        vx = msg.linear.x  # m/s
        wz = msg.angular.z  # rad/s
        if -0.2 < wz < 0.2:
            wz = 0.0
        vel_data = {"T": 13, "X": vx, "Z": wz}
        self.base.send_command(vel_data)

    def lights_callback(self, msg):
        if len(msg.data) != 2:
            self.get_logger().error(
                "LED control message must contain exactly two elements."
            )
            return
        IO4 = msg.data[0]
        IO5 = msg.data[1]
        led_ctrl_data = {
            "T": 132,
            "IO4": IO4,
            "IO5": IO5,
        }
        if self.base:
            self.base.send_command(led_ctrl_data)

    def servos_callback(self, msg):
        if len(msg.data) != 2:
            self.get_logger().error(
                "Servos control message must contain exactly two elements."
            )
            return
        yaw_deg = msg.data[0]
        pitch_deg = msg.data[1]
        servo_ctrl_data = {
            "T": 134,
            "X": yaw_deg,
            "Y": pitch_deg,
            "SX": 600,
            "SY": 600,
        }
        if self.base:
            self.base.send_command(servo_ctrl_data)

    def destroy_node(self):
        if self.base:
            self.vel_callback(Twist())  # send zero velocity command
            self.lights_callback(Float32MultiArray(data=[0.0, 0.0]))  # turn off lights
            self.servos_callback(Float32MultiArray(data=[0.0, 0.0]))  # reset servos
            self.base.destroy_base()
        self.get_logger().info("Serial Node has been shut down.")
        super().destroy_node()


def main():
    rclpy.init()
    node = SerialNode()
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException) as e:
        node.get_logger().warn(f"Serial node interrupted: {e}")
    finally:
        node.destroy_node()
        if ok():
            rclpy.shutdown()

File: ros2_ws/src/ugv_bringup_cpp/src/serial_node.cpp
#include <chrono>
#include <cmath>
#include <functional>
#include <fstream>
#include <memory>
#include <string>
#include <vector>

#include <nlohmann/json.hpp>

#include <ament_index_cpp/get_package_share_directory.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/imu.hpp>
#include <sensor_msgs/msg/joint_state.hpp>
#include <sensor_msgs/msg/magnetic_field.hpp>
#include <std_msgs/msg/float32.hpp>
#include <std_msgs/msg/float32_multi_array.hpp>

#include "ugv_bringup_cpp/base_controller.hpp"

namespace ugv_bringup_cpp {

class SerialNode : public rclcpp::Node {
public:
  SerialNode() : Node("serial_node") {
    baudrate_ = this->declare_parameter<int>("baudrate", 115200);
    port_ = this->declare_parameter<std::string>("port", "/dev/ttyTHS1");
    pub_name_imu_ = this->declare_parameter<std::string>("pub_name_imu", "imu/data_raw");
    pub_name_mag_ = this->declare_parameter<std::string>("pub_name_mag", "imu/mag");
    pub_name_odomraw_ = this->declare_parameter<std::string>("pub_name_odomraw", "encoder");
    pub_name_vol_ = this->declare_parameter<std::string>("pub_name_vol", "voltage");
    pub_name_joints_ = this->declare_parameter<std::string>("pub_name_joints", "joint_states");
    sub_name_vel_ = this->declare_parameter<std::string>("sub_name_vel", "cmd_vel");
    sub_name_lights_ = this->declare_parameter<std::string>("sub_name_lights", "ugv/led_strl");
    sub_name_servos_ = this->declare_parameter<std::string>("sub_name_servos", "ugv/servos");
    do_servo_calib_ = this->declare_parameter<bool>("do_servo_calib", true);
    calib_timeout_sec_ = this->declare_parameter<double>("calib_timeout_sec", 10.0);
    sample_period_ms_ = static_cast<double>(
        this->declare_parameter<int>("sample_period_ms", 50));

    if (baudrate_ <= 0) {
      baudrate_ = 115200;
    }
    if (calib_timeout_sec_ <= 0.0) {
      calib_timeout_sec_ = 10.0;
    }
    if (sample_period_ms_ <= 0.0) {
      sample_period_ms_ = 50.0;
    }

    base_ = std::make_unique<BaseController>(port_, baudrate_, this->get_logger());
    if (base_->IsOpen()) {
      base_->SendCommand(nlohmann::json{{"T", 131}, {"cmd", 1}});
      base_->SendCommand(nlohmann::json{{"T", 142}, {"cmd", sample_period_ms_}});
    }

    imu_acce_cov_.fill(0.0);
    imu_gyro_cov_.fill(0.0);
    mag_cov_.fill(0.0);
    LoadCovariances();

    if (do_servo_calib_) {
      StartServoCalib();
    }

    pub_imu_ = this->create_publisher<sensor_msgs::msg::Imu>(pub_name_imu_, 10);
    pub_mag_ = this->create_publisher<sensor_msgs::msg::MagneticField>(pub_name_mag_, 10);
    pub_odomraw_ = this->create_publisher<std_msgs::msg::Float32MultiArray>(pub_name_odomraw_, 10);
    pub_vol_ = this->create_publisher<std_msgs::msg::Float32>(pub_name_vol_, 10);
    pub_joints_ = this->create_publisher<sensor_msgs::msg::JointState>(pub_name_joints_, 10);

    timer_ = this->create_wall_timer(
        std::chrono::milliseconds(static_cast<int>(sample_period_ms_ / 2.0)),
        std::bind(&SerialNode::TimerCallback, this));

    sub_vel_ = this->create_subscription<geometry_msgs::msg::Twist>(
        sub_name_vel_, 1, std::bind(&SerialNode::VelCallback, this, std::placeholders::_1));
    sub_lights_ = this->create_subscription<std_msgs::msg::Float32MultiArray>(
        sub_name_lights_, 1, std::bind(&SerialNode::LightsCallback, this, std::placeholders::_1));
    sub_servos_ = this->create_subscription<std_msgs::msg::Float32MultiArray>(
        sub_name_servos_, 1, std::bind(&SerialNode::ServosCallback, this, std::placeholders::_1));

    RCLCPP_INFO(this->get_logger(), "Serial Node has been started.");
  }

  ~SerialNode() {
    if (base_) {
      base_->Stop();
    }
  }

private:
  void LoadCovariances() {
    try {
      std::string cov_path = ament_index_cpp::get_package_share_directory("ugv_bringup_cpp") +
                             std::string("/config/covariances.json");
      std::ifstream cov_file(cov_path);
      if (!cov_file.is_open()) {
        RCLCPP_ERROR(this->get_logger(), "Failed to open covariances file: %s", cov_path.c_str());
        return;
      }
      nlohmann::json covs;
      cov_file >> covs;
      if (covs.contains("imu_acce_covariance")) {
        FillCovariance(covs["imu_acce_covariance"], imu_acce_cov_);
      }
      if (covs.contains("imu_gyro_covariance")) {
        FillCovariance(covs["imu_gyro_covariance"], imu_gyro_cov_);
      }
      if (covs.contains("magnetic_field_covariance")) {
        FillCovariance(covs["magnetic_field_covariance"], mag_cov_);
      }
    } catch (const std::exception &e) {
      RCLCPP_ERROR(this->get_logger(), "Failed to load covariances: %s", e.what());
    }
  }

  void StartServoCalib() {
    if (!base_ || !base_->IsOpen()) {
      RCLCPP_WARN(this->get_logger(), "Skipping servo calibration: serial not available.");
      return;
    }
    if (calib_pending_) {
      return;
    }
    RCLCPP_INFO(this->get_logger(),
                "Starting servo calibration. Center servos now; will finalize after timeout.");
    base_->SendCommand(nlohmann::json{{"T", 210}, {"cmd", 0}});
    calib_pending_ = true;
    calib_timer_ = this->create_wall_timer(
        std::chrono::duration<double>(calib_timeout_sec_),
        std::bind(&SerialNode::FinishServoCalib, this));
  }

  void FinishServoCalib() {
    if (!base_ || !base_->IsOpen()) {
      RCLCPP_WARN(this->get_logger(), "Skipping servo calibration finalize: serial not available.");
      return;
    }
    if (!calib_pending_) {
      return;
    }
    base_->SendCommand(nlohmann::json{{"T", 502}, {"id", 1}});
    base_->SendCommand(nlohmann::json{{"T", 502}, {"id", 2}});
    calib_pending_ = false;
    if (calib_timer_) {
      calib_timer_->cancel();
    }
    RCLCPP_INFO(this->get_logger(), "Servo calibration completed.");
  }

  void TimerCallback() {
    if (!base_ || !base_->IsOpen()) {
      return;
    }
    auto base_data_opt = base_->FeedbackData();
    if (!base_data_opt.has_value()) {
      return;
    }
    const auto &base_data = base_data_opt.value();
    if (!base_data.contains("T") || !base_data["T"].is_number_integer() ||
        base_data["T"].get<int>() != 1001) {
      return;
    }
    auto time_now = this->now();
    PublishImu(time_now, base_data);
    PublishMag(time_now, base_data);
    PublishOdomRaw(base_data);
    PublishVoltage(base_data);
    PublishJoints(time_now, base_data);
  }

  void PublishImu(const rclcpp::Time &time_now, const nlohmann::json &base_data) {
    sensor_msgs::msg::Imu imu_msg;
    imu_msg.header.stamp = time_now;
    imu_msg.header.frame_id = "base_imu_link";
    imu_msg.linear_acceleration.x = 9.8 * base_data.value("ax", 0.0) / 8192.0;
    imu_msg.linear_acceleration.y = 9.8 * base_data.value("ay", 0.0) / 8192.0;
    imu_msg.linear_acceleration.z = 9.8 * base_data.value("az", 0.0) / 8192.0;
    imu_msg.linear_acceleration_covariance = imu_acce_cov_;
    imu_msg.angular_velocity.x = ToRad(base_data.value("gx", 0.0) / 16.4);
    imu_msg.angular_velocity.y = ToRad(base_data.value("gy", 0.0) / 16.4);
    imu_msg.angular_velocity.z = ToRad(base_data.value("gz", 0.0) / 16.4);
    imu_msg.angular_velocity_covariance = imu_gyro_cov_;
    imu_msg.orientation_covariance[0] = -1;
    pub_imu_->publish(imu_msg);
  }

  void PublishMag(const rclcpp::Time &time_now, const nlohmann::json &base_data) {
    sensor_msgs::msg::MagneticField mag_msg;
    mag_msg.header.stamp = time_now;
    mag_msg.header.frame_id = "base_imu_link";
    mag_msg.magnetic_field.x = base_data.value("mx", 0.0) * 0.15 * 1e-7;
    mag_msg.magnetic_field.y = base_data.value("my", 0.0) * 0.15 * 1e-7;
    mag_msg.magnetic_field.z = base_data.value("mz", 0.0) * 0.15 * 1e-7;
    mag_msg.magnetic_field_covariance = mag_cov_;
    pub_mag_->publish(mag_msg);
  }

  void PublishOdomRaw(const nlohmann::json &base_data) {
    std_msgs::msg::Float32MultiArray msg;
    msg.data = {
        static_cast<float>(base_data.value("odl", 0.0) / 100.0),
        static_cast<float>(base_data.value("odr", 0.0) / 100.0),
        static_cast<float>(base_data.value("L", 0.0)),
        static_cast<float>(base_data.value("R", 0.0)),
    };
    pub_odomraw_->publish(msg);
  }

  void PublishVoltage(const nlohmann::json &base_data) {
    std_msgs::msg::Float32 msg;
    msg.data = static_cast<float>(base_data.value("v", 0.0) / 100.0);
    pub_vol_->publish(msg);
  }

  void PublishJoints(const rclcpp::Time &time_now, const nlohmann::json &base_data) {
    sensor_msgs::msg::JointState msg;
    msg.header.stamp = time_now;
    msg.name = {"pt_base_link_to_pt_link1", "pt_link1_to_pt_link2"};
    msg.position = {ToRad(base_data.value("pan", 0.0)), ToRad(base_data.value("tilt", 0.0))};
    pub_joints_->publish(msg);
  }

  void VelCallback(const geometry_msgs::msg::Twist::SharedPtr msg) {
    if (!base_ || !base_->IsOpen()) {
      return;
    }
    double vx = msg->linear.x;
    double wz = msg->angular.z;
    if (wz > -0.2 && wz < 0.2) {
      wz = 0.0;
    }
    base_->SendCommand(nlohmann::json{{"T", 13}, {"X", vx}, {"Z", wz}});
  }

  void LightsCallback(const std_msgs::msg::Float32MultiArray::SharedPtr msg) {
    if (msg->data.size() != 2) {
      RCLCPP_ERROR(this->get_logger(), "LED control message must contain exactly two elements.");
      return;
    }
    if (!base_ || !base_->IsOpen()) {
      return;
    }
    base_->SendCommand(nlohmann::json{{"T", 132}, {"IO4", msg->data[0]}, {"IO5", msg->data[1]}});
  }

  void ServosCallback(const std_msgs::msg::Float32MultiArray::SharedPtr msg) {
    if (msg->data.size() != 2) {
      RCLCPP_ERROR(this->get_logger(), "Servos control message must contain exactly two elements.");
      return;
    }
    if (!base_ || !base_->IsOpen()) {
      return;
    }
    base_->SendCommand(nlohmann::json{{"T", 134}, {"X", msg->data[0]}, {"Y", msg->data[1]},
                                      {"SX", 600}, {"SY", 600}});
  }

  static constexpr double kPi = 3.14159265358979323846;
  static double ToRad(double degrees) { return degrees * kPi / 180.0; }

  static void FillCovariance(const nlohmann::json &json_array,
                             std::array<double, 9> &target) {
    if (!json_array.is_array() || json_array.size() != target.size()) {
      return;
    }
    for (std::size_t i = 0; i < target.size(); ++i) {
      if (json_array[i].is_number()) {
        target[i] = json_array[i].get<double>();
      }
    }
  }

  int baudrate_{115200};
  std::string port_;
  std::string pub_name_imu_;
  std::string pub_name_mag_;
  std::string pub_name_odomraw_;
  std::string pub_name_vol_;
  std::string pub_name_joints_;
  std::string sub_name_vel_;
  std::string sub_name_lights_;
  std::string sub_name_servos_;
  bool do_servo_calib_{true};
  double calib_timeout_sec_{10.0};
  double sample_period_ms_{50.0};

  std::unique_ptr<BaseController> base_;
  bool calib_pending_{false};
  rclcpp::TimerBase::SharedPtr calib_timer_;
  rclcpp::TimerBase::SharedPtr timer_;

  rclcpp::Publisher<sensor_msgs::msg::Imu>::SharedPtr pub_imu_;
  rclcpp::Publisher<sensor_msgs::msg::MagneticField>::SharedPtr pub_mag_;
  rclcpp::Publisher<std_msgs::msg::Float32MultiArray>::SharedPtr pub_odomraw_;
  rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr pub_vol_;
  rclcpp::Publisher<sensor_msgs::msg::JointState>::SharedPtr pub_joints_;
  rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr sub_vel_;
  rclcpp::Subscription<std_msgs::msg::Float32MultiArray>::SharedPtr sub_lights_;
  rclcpp::Subscription<std_msgs::msg::Float32MultiArray>::SharedPtr sub_servos_;

  std::array<double, 9> imu_acce_cov_{};
  std::array<double, 9> imu_gyro_cov_{};
  std::array<double, 9> mag_cov_{};
};

}  // namespace ugv_bringup_cpp

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<ugv_bringup_cpp::SerialNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
